### ISSUE TITLE
Strange cursor behaviour

### ISSUE THREAD

**Issue by pallinp** [oracle/node-oracledb#239]
Hi,

I'm currently facing a problem of opened cursor using oracledb for nodejs.

In fact, when a query fail (unique constraint fail for example) it looks like the query statement is not free and for each error I have +1 opened cursor in my db.

Strangely Ive got no cursor issues when queries succeed.

Here is my small encapsulation of the driver:
```js
'use strict';

var oracleDb = require( 'oracledb' );


function OracleManager() {
    /* Nothing to do here */
}

OracleManager.prototype = {
    constructor:OracleManager,

    connect:function( env, callback ) {
        var _ctrl = this;

        oracleDb.outFormat = oracleDb.OBJECT;
        oracleDb.createPool( {
            user          : env.settings.oracle.user,
            password      : env.settings.oracle.password,
            connectString : env.settings.oracle.connectString,
            poolMax       : 1,
            poolMin       : 0,
            poolIncrement : 1,
            poolTimeout   : 0,
            stmtCacheSize : 0
        }, function( error, pool ) {
            if ( error ) {
                console.error( error.message );
                return ;
            }
            _ctrl.pool = pool;
            callback();
        } );
    },

    executeQuery:function( env, query, callback ) {
        var _ctrl = this;

        var time = Date.now();
        this.pool.getConnection( function( error, connection ) {
            if ( error ) {
                console.error( error.message );
                callback( null );
                return ;
            }
            if ( query.type === 'SELECT' ) {
                _ctrl.executeSelect( connection, env, query, callback );
            } else {
                _ctrl.executeInsert( connection, env, query, callback );
            }
        } );
    },

    executeInsert:function( connection, env, query, callback ) {
        var _ctrl = this;

        connection.execute(
            query.string,
            query.args,
            {
                autoCommit : true
            },
            function( error, result ) {
                if ( error ) {
                    console.error( error.message );
                    _ctrl.releaseConnection( connection, function() {
                        callback( null );
                    } );
                    return ;
                }
                _ctrl.releaseConnection( connection, function() {
                    callback( result );
                } );
            } );
    },

    executeSelect:function( connection, env, query, callback ) {
        var _ctrl = this;

        connection.execute(
            query.string,
            query.args,
            {
                resultSet : true
            },
            function( error, result ) {
                if ( error ) {
                    console.error( error.message );
                    _ctrl.releaseConnection( connection, function() {
                        callback( null );
                    } );
                    return ;
                }
                result.rows = [];
                _ctrl.fetchRows( connection, env, result, callback );
            }
        );
    },

    releaseConnection:function( connection, callback ) {
        connection.release( function( error ) {
            if ( error ) {
                console.error( error.message );
            }
            callback();
        } );
    },

    fetchRows:function( connection, env, result, callback ) {
        var _ctrl = this;

        result.resultSet.getRows(
            1000000,
            function( error, rows ) {
                if ( error ) {
                    result.resultSet.close( function() {
                        console.error( error.message );
                        _ctrl.releaseConnection( connection, function() {
                            callback( null );
                        } );
                    } );
                } else if ( rows.length === 0 ) {
                    result.resultSet.close( function() {
                        _ctrl.releaseConnection( connection, function() {
                            callback( result );
                        } );
                    } );
                } else {
                    result.rows = result.rows.concat( rows );
                    _ctrl.fetchRows( connection, env, result, callback );
                }
            }
        );
    },

    close:function() {
        this.pool.terminate( function( error ) {
            if ( error ) {
                console.error( error.message );
            }
        } );
    },

};

module.exports = OracleManager;
```

And there is the tested code:

```js
'use strict';

var dbObject = require( './OracleManager.js' );
var async    = require( 'async' );

var db = new dbObject();
var env = {
    settings : {
        oracle : {
            user : xxxxxxxxx,
            password : xxxxxxxx,
            connectString : xxxxxxxx
        }
    }
};

var time = Date.now();

db.connect( env, function() {
    
    console.log( ( ( Date.now() - time ) / 1000 ) + 's' );

  var query = {
            string : 'SELECT * FROM kjizeofiofjioz',
            args : [],
            type : 'SELECT'
        };

    var queries = [];
    for ( var i = 0; i < 10000; ++i ) {
        queries.push( query );
    }

    var tasks = async.queue( function( item, next ) {
        db.executeQuery( env, item, function( result ) {
            next();
        } );
    } );

    tasks.push( queries );

} );
```

Obviously the table 'kjizeofiofjioz' does not exist and my openend cursor grow up very fast...

Regards,

pallinp

**Comment by cjbj** [comment]
We'll look at it.

**Comment by pallinp** [comment]
Hi,

Do you have some news about this issue ?

Regards,

pallinp

**Comment by cjbj** [comment]
It's a bug. The cause is the OCI statement handle not being released.

**Comment by cjbj** [comment]
@pallinp this is fixed in node-oracledb 1.4, which has just been released.
